pkg/controllers/provisioning/provisioner.go

```diff
@@ -249,6 +247,26 @@ func (p *Provisioner) bind(ctx context.Context, node *v1.Node, pods []*v1.Pod) (
 	return nil
 }
 
+func (p *Provisioner) bind(ctx context.Context, node *v1.Node, pod *v1.Pod) error {
+	if err := p.coreV1Client.Pods(pod.Namespace).Bind(ctx, &v1.Binding{TypeMeta: pod.TypeMeta, ObjectMeta: pod.ObjectMeta, Target: v1.ObjectReference{Name: node.Name}}, metav1.CreateOptions{}); err != nil {
+		return fmt.Errorf("binding pod %s/%s, %w", pod.Name, pod.Namespace, err)
+	}
+	for _, volume := range pod.Spec.Volumes {
+		if volume.PersistentVolumeClaim == nil {
+			continue
+		}
+		pvc := &v1.PersistentVolumeClaim{}
+		if err := p.kubeClient.Get(ctx, types.NamespacedName{Name: volume.PersistentVolumeClaim.ClaimName, Namespace: pod.Namespace}, pvc); err != nil {
+			return fmt.Errorf("getting persistent volume claim %s, %w", volume.PersistentVolumeClaim.ClaimName, err)
+		}
+		pvc.Annotations["volume.kubernetes.io/selected-node"] = node.Name
```

Comment:
I think it is necessary for some routine to wait for every `PVC provision` to succeed/fail after setting this annotation to trigger `PVC provision`. Step 3 here: https://github.com/kubernetes/design-proposals-archive/blob/main/storage/volume-topology-scheduling.md#bind.

Because if a `PVC provision` fails then the CSI driver's provisioner controller signals failure by removing the selected-node annotation https://github.com/kubernetes-sigs/sig-storage-lib-external-provisioner/blob/a35bec6eb688ef7a0e31b340ed77bdb84010ace7/controller/controller.go#L1298. And some routine has to notice that + add the annotation again to trigger another `PVC provision`.

The default scheduler prebind implementation of this wait routine is: https://github.com/kubernetes/kubernetes/blob/39c76ba2edeadb84a115cc3fbd9204a2177f1c28/pkg/scheduler/framework/plugins/volumebinding/binder.go#L462-L465

Reply:
Interesting. I may just implement a PVC scheduler that watches PVCs and pods and makes sure that the annotation is set for the node. Thanks for pointing this out.